lsp4clj/src/clojure_lsp/lsp.clj

```diff
@@ -0,0 +1,436 @@
+    (start :semanticTokensFull
+           (async-request params handler/semantic-tokens-full handler ::coercer/semantic-tokens)))
+
+  (^CompletableFuture semanticTokensRange [_ ^SemanticTokensRangeParams params]
+    (start :semanticTokensRange
+           (async-request params handler/semantic-tokens-range handler ::coercer/semantic-tokens)))
+
+  (^CompletableFuture prepareCallHierarchy [_ ^CallHierarchyPrepareParams params]
+    (start :prepareCallHierarchy
+           (async-request params handler/prepare-call-hierarchy handler ::coercer/call-hierarchy-items)))
+
+  (^CompletableFuture callHierarchyIncomingCalls [_ ^CallHierarchyIncomingCallsParams params]
+    (start :callHierarchyIncomingCalls
+           (async-request params handler/call-hierarchy-incoming handler ::coercer/call-hierarchy-incoming-calls)))
+
+  (^CompletableFuture callHierarchyOutgoingCalls [_ ^CallHierarchyOutgoingCallsParams params]
+    (start :callHierarchyOutgoingCalls
+           (async-request params handler/call-hierarchy-outgoing handler ::coercer/call-hierarchy-outgoing-calls)))
+
+  (^CompletableFuture linkedEditingRange [_ ^LinkedEditingRangeParams params]
+    (start :linkedEditingRange
+           (async-request params handler/linked-editing-ranges handler ::coercer/linked-editing-ranges-or-error))))
+
+(deftype LSPWorkspaceService [handler]
+  WorkspaceService
+  (^CompletableFuture executeCommand [_ ^ExecuteCommandParams params]
+    (start :executeCommand
+           (future
+             (sync-notification params handler/execute-command handler)))
+    (CompletableFuture/completedFuture 0))
+
+  (^void didChangeConfiguration [_ ^DidChangeConfigurationParams params]
+    (log/warn (coercer/java->clj params)))
+
+  (^void didChangeWatchedFiles [_ ^DidChangeWatchedFilesParams params]
+    (start :didChangeWatchedFiles
+           (async-notification params handler/did-change-watched-files handler)))
+
+  ;; TODO wait for lsp4j release
+  #_(^void didDeleteFiles [_ ^DeleteFilesParams params]
+                          (start :didSave
+                                 (sync-handler params handler/did-delete-files habdler)))
+
+  (^CompletableFuture symbol [_ ^WorkspaceSymbolParams params]
+    (start :workspaceSymbol
+           (async-request params handler/workspace-symbols handler ::coercer/workspace-symbols))))
+
+(defn client-capabilities [^InitializeParams params]
+  (some->> params
+           .getCapabilities
+           (coercer/conform-or-log ::coercer/client-capabilities)))
+
+(defn start-parent-process-liveness-probe!
+  [ppid server]
+  (go-loop []
+    (<! (timeout 5000))
+    (if (shared/process-alive? ppid)
+      (recur)
+      (do
+        (log/info "Parent process" ppid "is not running - exiting server")
+        (.exit ^LanguageServer server)))))
+
+#_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
+(deftype LSPServer [handler db capabilities client-settings]
```

Comment:
We are probably having reflection issues here that will affect graalvm since this code is kind of generic on some parts (the macro ones), so it's a good idea to type hint `handler` here and on the other deftypes.

```suggestion
(deftype LSPServer [^IHandler handler db capabilities client-settings]
```